PowerEditor/src/WinControls/OpenSaveFileDialog/CustomFileDialog.cpp

```diff
@@ -625,7 +649,37 @@ class FileDialogEventHandler : public IFileDialogEvents, public IFileDialogContr
 				HWND hwnd = GetFocus();
 				auto it = s_handleMap.find(hwnd);
 				if (it != s_handleMap.end() && it->second && hwnd == it->second->_hwndNameEdit)
-					it->second->onPreFileOk();
+				{
+					it->second->_isHangul = IsHangul(it->second->_keyboardLayoutLanguage);
+
+					if (it->second->_isHangul)
+					{
+						// If IME is in Hangul mode, ignore VK_RETURN to complete the composition and change to alphabetical input mode to proceed to onPreFileOk() on the next VK_RETURN.
+						auto himc = ImmGetContext(hwnd);
+						ImmSetConversionStatus(himc, IME_CMODE_ALPHANUMERIC, IME_SMODE_NONE);
+						ImmReleaseContext(hwnd, himc);
+					}
+					else
+					{
+						it->second->onPreFileOk();
+					}
+				}
+			}
+		}
+		return ::CallNextHookEx(nullptr, nCode, wParam, lParam);
+	}
+
+	// Dectect language layout of keyboard
+	static LRESULT CALLBACK LanguageDetectHook(int nCode, WPARAM wParam, LPARAM lParam)
+	{
+		if (nCode == HSHELL_LANGUAGE)
+		{
+			HWND hwnd = GetFocus();
+			auto it = s_handleMap.find(hwnd);
+			if (it != s_handleMap.end() && it->second && hwnd == it->second->_hwndNameEdit)
+			{
+				HKL hkl = (HKL)lParam;
+				it->second->_keyboardLayoutLanguage = PRIMARYLANGID(LOWORD(hkl));
```

Comment:
After testing, this line has never been run, obviously the condition `if (it != s_handleMap.end() && it->second && hwnd == it->second->_hwndNameEdit)` is always false.
Any idea about this issue?

Reply:
That section of code won't be called unless the user changes the input language.   
Currently, _keyboardLayoutLanguage is only checking to see if it's Korean, so I didn't initialize it.  
Do you think it would be better for _keyboardLayoutLanguage to be initialized to English or the system locale? Or any other ideas?

Reply:
> That section of code won't be called unless the user changes the input language.

How come the condition `if (it != s_handleMap.end() && it->second && hwnd == it->second->_hwndNameEdit)` will turn into true when user changes the input language? 
And BTW, I add the 2 Chinese keyboard layouts into my system, by switching into them it didn't change anything. 

> Do you think it would be better for _keyboardLayoutLanguage to be initialized to English or the system locale? Or any other ideas?

It will be the next step. But for the moment, I just can't see how and why the condition mentioned above can be true.
Please provide the instruction to make line `it->second->_keyboardLayoutLanguage = PRIMARYLANGID(LOWORD(hkl));` to be run.

Reply:
Let me correct: when the user changes input language, ```if (nCode == HSHELL_LANGUAGE)``` becomes true. ```if (it != s_handleMap.end() && it->second && hwnd == it->second->_hwndNameEdit)``` is looking for a ```FileDialogEventHandler``` like the other hooking functions above to access ```_keyboardLayoutLanguage```. I don't understand what's wrong with that code, are you experiencing the same issue with the other hooking function above?

Reply:
Strangely now I can trigger this condition to true by switching ON my Chinese input...
I guess it's due to startup of Windows.
Thank you for the explanation.

Reply:
But still, if my Chinese input is already enabled, and I trigger save dialog, _keyboardLayoutLanguage is 0.
So let's detect input language while dialog is launched and initialize  _keyboardLayoutLanguage correctely.